File: solutions/javascript/dfs/numberOfIslands.js
const LAND = '1';

const getPositionString = (row, col) => `${row}, ${col}`;
const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
];

const isInBounds = (grid, row, col) => {
    const numRows = grid.length;
    const numCols = grid[0].length;

    const rowInBounds = row >= 0 && row < numRows;
    const colInBounds = col >= 0 && col < numCols;

    return rowInBounds && colInBounds;
};

const markIslandAsVisited = (grid, row, col, visited) => {
    // Base case
    if (!isInBounds(grid, row, col)) return;

    const curPositionString = getPositionString(row, col);
    if (visited.has(curPositionString)) return;

    if (grid[row][col] !== LAND) return;

    // Process node
    visited.add(curPositionString);

    // Recurse on potential neighbors
    for (const direction of directions) {
        const [rowChange, colChange] = direction;

        const newRow = row + rowChange;
        const newCol = col + colChange;

        markIslandAsVisited(grid, newRow, newCol, visited);
    }
};

const numIslands = (grid) => {
    const numRows = grid.length;
    const numCols = grid[0].length;

    let numberOfIslands = 0;
    const visited = new Set();

    for (let row = 0; row < numRows; row++) {
        for (let col = 0; col < numCols; col++) {
            const curPositionString = getPositionString(row, col);

            if (visited.has(curPositionString)) continue;

            const terrainType = grid[row][col];
            if (terrainType !== LAND) continue;

            numberOfIslands++;
            markIslandAsVisited(grid, row, col, visited);
        }
    }

    return numberOfIslands;
};

File: solutions/python/dfs/number_of_islands.py
LAND = '1'
directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
]

def is_in_bounds(grid, row, col):
    num_rows = len(grid)
    num_cols = len(grid[0])

    row_in_bounds = row >= 0 and row < num_rows
    col_in_bounds = col >= 0 and col < num_cols

    return row_in_bounds and col_in_bounds

class Solution:
    def mark_island_as_visited(self, grid, row, col, visited):
        # Base case
        if not is_in_bounds(grid, row, col):
            return

        if (row, col) in visited:
            return

        if grid[row][col] != LAND:
            return

        # Process node
        visited.add((row, col))

        # Recurse on potential neighbors
        for direction in directions:
            row_change, col_change = direction

            new_row = row + row_change
            new_col = col + col_change

            self.mark_island_as_visited(grid, new_row, new_col, visited)

    def numIslands(self, grid: List[List[str]]) -> int:
        num_rows = len(grid)
        num_cols = len(grid[0])

        number_of_islands = 0
        visited = set()

        for row in range(num_rows):
            for col in range(num_cols):
                if (row, col) in visited:
                    continue

                terrain_type = grid[row][col]
                if terrain_type != LAND: continue

                number_of_islands += 1
                self.mark_island_as_visited(grid, row, col, visited)

        return number_of_islands